kimchi/src/alphas.rs

```diff
@@ -0,0 +1,276 @@
+//!
+//! - when creating the index, we do not know alpha at this point so we
+//!   simply keep track of what constraints will use what powers
+//! - when creating a proof or verifying a proof, at this point we know alpha
+//!   so we can use the mapping we created during the creation of the index.
+//!
+//! For this to work, we use two types:
+//!
+//! - [Builder], which allows us to map constraints to powers
+//! - [Alphas], which you can derive from [Builder] and an `alpha`
+//!
+//! Both construction will enforce that you use all the powers of
+//! alphas that you register for constraint. This allows us to
+//! make sure that we compute the correct amounts, without reusing
+//! powers of alphas between constraints.
+
+use ark_ff::Field;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    iter::{Cloned, Take},
+    ops::Range,
+    slice::Iter,
+    thread,
+};
+
+// ------------------------------------------
+
+/// A constraint type represents a polynomial that will be part of the final equation f (the circuit equation)
+#[derive(PartialEq, Eq, Clone, Hash, Debug, Serialize, Deserialize)]
+pub enum ConstraintType {
+    /// gates in the PLONK constraint system.
+    /// As gates are mutually exclusive (only a single selector polynomial
+    /// will be non-zero for a given point),
+    /// we can reuse the same power of alphas accross gates.
+    Gate,
+    /// The permutation argument
+    Permutation,
+    /// The lookup argument
+    Lookup,
+}
+
+// ------------------------------------------
+
+/// This type can be used to create a mapping between powers of alpha and constraint types.
+/// See [Builder::default] to create one,
+/// and [Builder::register] to register a new mapping.
+/// Once you know the alpha value, you can convert this type to a [Alphas].
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+pub struct Builder {
+    /// The next power of alpha to use
+    next_power: usize,
+    /// The mapping between constraint types and powers of alpha
+    mapping: HashMap<ConstraintType, Range<usize>>,
+}
+
+impl Builder {
+    /// Registers a new [ConstraintType],
+    /// associating it a number `powers` of powers of alpha.
+    /// The function returns an iterator of powers that must be used.
+    #[must_use]
+    pub fn register(&mut self, ty: ConstraintType, powers: usize) -> Range<usize> {
+        let new_power = self.next_power + powers;
+        let range = self.next_power..new_power;
+        if self.mapping.insert(ty.clone(), range.clone()).is_some() {
+            panic!("you are attempting to register {:?} twice", ty);
+        }
+        self.next_power = new_power;
+        range
+    }
+
+    /// Creates a new instance of [Alphas] via a [Builder] and value `alpha`.
+    pub fn into_alphas<F: Field>(self, alpha: F) -> Alphas<F> {
+        let mut last_power = F::one();
+        let mut alphas = Vec::with_capacity(self.next_power);
+        alphas.push(F::one());
```

Comment:
you're right! There is an off-by-one here, I pre-compute one unnecessary power of alpha

Reply:
Ah, "pre-compute", but not actually using it for the range... Cool, in any case I guess it is fine to avoid the unnecessary additional one (which I think you already fixed in the new commit 👍 )